java/src/main/native/src/row_conversion.cu

```diff
@@ -304,8 +329,620 @@ copy_from_fixed_width_columns(const cudf::size_type start_row, const cudf::size_
+                             const size_type *col_sizes, const size_type *col_offsets,
+                             const size_type *row_offsets, int8_t **output_data) {
+  // We are going to copy the data in two passes.
+  // The first pass copies a chunk of data into shared memory.
+  // The second pass copies that chunk from shared memory out to the final location.
+
+  // Because shared memory is limited we copy a subset of the rows at a time.
+  // This has been broken up for us in the block_info struct, so we don't have
+  // any calculation to do here, but it is important to note.
+
+  constexpr unsigned stages_count = NUM_BLOCKS_PER_KERNEL_LOADED;
+  auto group = cooperative_groups::this_thread_block();
+  extern __shared__ int8_t shared_data[];
+  int8_t *shared[stages_count] = {shared_data, shared_data + shmem_used_per_block};
+
+  __shared__ cuda::barrier<cuda::thread_scope_block> block_barrier[NUM_BLOCKS_PER_KERNEL_LOADED];
+  if (group.thread_rank() == 0) {
+    for (int i = 0; i < NUM_BLOCKS_PER_KERNEL_LOADED; ++i) {
+      init(&block_barrier[i], group.size());
+    }
+  }
+
+  group.sync();
+
+  auto const blocks_remaining =
+      std::min((uint)block_infos.size() - blockIdx.x * NUM_BLOCKS_PER_KERNEL_TO_ROWS,
+               (uint)NUM_BLOCKS_PER_KERNEL_TO_ROWS);
+
+  size_t fetch;
+  size_t subset;
+  for (subset = fetch = 0; subset < blocks_remaining; ++subset) {
+    // Fetch ahead up to stages_count subsets
+    for (; fetch < blocks_remaining && fetch < (subset + stages_count); ++fetch) {
+      auto const fetch_block = block_infos[blockIdx.x * NUM_BLOCKS_PER_KERNEL_TO_ROWS + fetch];
+      auto const num_fetch_cols = fetch_block.num_cols();
+      auto const num_fetch_rows = fetch_block.num_rows();
+      auto const num_elements_in_block = num_fetch_cols * num_fetch_rows;
+      auto const fetch_block_row_size = fetch_block.get_shared_row_size(col_offsets, col_sizes);
+      auto const starting_column_offset = col_offsets[fetch_block.start_col];
+      auto &fetch_barrier = block_barrier[fetch % NUM_BLOCKS_PER_KERNEL_LOADED];
+
+      // wait for the last use of the memory to be completed
+      if (fetch >= NUM_BLOCKS_PER_KERNEL_LOADED) {
+        fetch_barrier.arrive_and_wait();
+      }
+
+      // to do the copy we need to do n column copies followed by m element copies OR
+      // we have to do m element copies followed by r row copies. When going from column
+      // to row it is much easier to copy by elements first otherwise we would need a running
+      // total of the column sizes for our block, which isn't readily available. This makes it
+      // more appealing to copy element-wise from input data into shared matching the end layout
+      // and do row-based memcopies out.
+
+      auto const shared_buffer_base = shared[fetch % stages_count];
+      for (auto el = (int)threadIdx.x; el < num_elements_in_block; el += blockDim.x) {
```

Comment:
```suggestion
      for (auto el = static_cast<int>(threadIdx.x); el < num_elements_in_block; el += blockDim.x) {
```

Should do a pass over this whole file and switch all the manual casting to static_cast.